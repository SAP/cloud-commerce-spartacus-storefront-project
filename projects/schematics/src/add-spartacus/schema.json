{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "baseUrl": {
      "type": "string",
      "description": "The url of the OCC backend",
      "default": "https://localhost:9002"
    },
    "baseSite": {
      "type": "string",
      "description": "The base site to use with spartacus",
      "default": "electronics"
    },
    "useMetaTags": {
      "type": "boolean",
      "description": "Configure baseUrl via meta tag in index.html"
    },
    "featureLevel": {
      "type": "string",
      "description": "Configure feature level for application.",
      "default": "1.1"
    },
    "overwriteAppComponent": {
      "type": "boolean",
      "description": "Overwrite whole app.component.html file.",
      "default": false
    }
  },
  "required": []
}
