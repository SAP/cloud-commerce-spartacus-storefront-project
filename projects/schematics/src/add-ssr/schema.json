{
  "$schema": "http://json-schema.org/schema",
  "id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    }
  },
  "required": []
}
