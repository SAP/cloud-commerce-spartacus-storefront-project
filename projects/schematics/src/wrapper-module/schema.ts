import { ExecutionOptions } from '@angular-devkit/schematics';

export interface Schema extends Partial<ExecutionOptions> {
  /**
   * The name of the project in which to execute schematics.
   */
  project: string;
  /**
   * When enabled, prints the additional logs.
   */
  debug?: boolean;
  /**
   * The marker module for which
   * to find or generate a new wrapper module.
   */
  markerModuleName: string;
  /**
   * Name of the feature module
   * to add to the wrapper module.
   */
  featureModuleName: string;
  /**
   * Internal options.
   * Should not be set by the user.
   */
  internal?: {
    /**
     * The execution sequence.
     * Used for ensuring one log message is printed.
     */
    sequence?: number;
  };
}
