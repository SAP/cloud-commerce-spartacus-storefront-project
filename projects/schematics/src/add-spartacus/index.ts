import { experimental, JsonParseMode, parseJson } from '@angular-devkit/core';
import { italic, red } from '@angular-devkit/core/src/terminal';
import {
  chain,
  noop,
  Rule,
  SchematicContext,
  SchematicsException,
  Tree,
} from '@angular-devkit/schematics';
import { NodePackageInstallTask } from '@angular-devkit/schematics/tasks';
import {
  appendHtmlElementToHead,
  getProjectStyleFile,
  getProjectTargetOptions,
} from '@angular/cdk/schematics';
import {
  addSymbolToNgModuleMetadata,
  insertImport,
  isImported,
} from '@schematics/angular/utility/ast-utils';
import { InsertChange } from '@schematics/angular/utility/change';
import {
  addPackageJsonDependency,
  NodeDependency,
  NodeDependencyType,
} from '@schematics/angular/utility/dependencies';
import { getAppModulePath } from '@schematics/angular/utility/ng-ast-utils';
import { getProjectTargets } from '@schematics/angular/utility/project-targets';
import { Schema as SpartacusOptions } from './schema';
import { getTsSourceFile } from '../shared/utils/file-utils';

function getWorkspace(
  host: Tree
): { path: string; workspace: experimental.workspace.WorkspaceSchema } {
  const possibleFiles = ['/angular.json', '/.angular.json'];
  const path = possibleFiles.filter(filePath => host.exists(filePath))[0];

  if (!path) {
    throw new SchematicsException(`Could not find Angular`);
  }

  const configBuffer = host.read(path);
  if (configBuffer === null) {
    throw new SchematicsException(`Could not find (${path})`);
  }
  const content = configBuffer.toString();

  return {
    path,
    workspace: (parseJson(
      content,
      JsonParseMode.Loose
    ) as {}) as experimental.workspace.WorkspaceSchema,
  };
}

function addPackageJsonDependencies(): Rule {
  const spartacusVersion = '^1.0.0';
  const ngrxVersion = '^8.0.0';

  return (tree: Tree, context: SchematicContext) => {
    const dependencies: NodeDependency[] = [
      {
        type: NodeDependencyType.Default,
        version: spartacusVersion,
        name: '@spartacus/core',
      },
      {
        type: NodeDependencyType.Default,
        version: spartacusVersion,
        name: '@spartacus/storefront',
      },
      {
        type: NodeDependencyType.Default,
        version: spartacusVersion,
        name: '@spartacus/assets',
      },
      {
        type: NodeDependencyType.Default,
        version: spartacusVersion,
        name: '@spartacus/styles',
      },

      {
        type: NodeDependencyType.Default,
        version: '^4.1.0',
        name: '@ng-bootstrap/ng-bootstrap',
      },
      {
        type: NodeDependencyType.Default,
        version: '^2.13.2',
        name: '@ng-select/ng-select',
      },

      {
        type: NodeDependencyType.Default,
        version: ngrxVersion,
        name: '@ngrx/store',
      },
      {
        type: NodeDependencyType.Default,
        version: ngrxVersion,
        name: '@ngrx/effects',
      },
      {
        type: NodeDependencyType.Default,
        version: ngrxVersion,
        name: '@ngrx/router-store',
      },

      {
        type: NodeDependencyType.Default,
        version: '^4.2.1',
        name: 'bootstrap',
      },
      { type: NodeDependencyType.Default, version: '^15.0.6', name: 'i18next' },
      {
        type: NodeDependencyType.Default,
        version: '^2.0.1',
        name: 'i18next-xhr-backend',
      },
    ];

    dependencies.forEach(dependency => {
      addPackageJsonDependency(tree, dependency);
      context.logger.log(
        'info',
        `✅️ Added '${dependency.name}' into ${dependency.type}`
      );
    });

    return tree;
  };
}

function installPackageJsonDependencies(): Rule {
  return (tree: Tree, context: SchematicContext) => {
    context.addTask(new NodePackageInstallTask());
    context.logger.log('info', `🔍 Installing packages...`);
    return tree;
  };
}

function addImport(
  host: Tree,
  modulePath: string,
  importText: string,
  importPath: string
) {
  const moduleSource = getTsSourceFile(host, modulePath) as any;
  if (!isImported(moduleSource, importText, importPath)) {
    const change = insertImport(
      moduleSource,
      modulePath,
      importText,
      importPath
    );
    if (change) {
      const recorder = host.beginUpdate(modulePath);
      recorder.insertLeft(
        (change as InsertChange).pos,
        (change as InsertChange).toAdd
      );
      host.commitUpdate(recorder);
    }
  }
}

function importModule(host: Tree, modulePath: string, importText: string) {
  const moduleSource = getTsSourceFile(host, modulePath);
  const metadataChanges = addSymbolToNgModuleMetadata(
    moduleSource,
    modulePath,
    'imports',
    importText
  );
  if (metadataChanges) {
    const recorder = host.beginUpdate(modulePath);
    metadataChanges.forEach((change: InsertChange) => {
      recorder.insertRight(change.pos, change.toAdd);
    });
    host.commitUpdate(recorder);
  }
}

function getStorefrontConfig(options: SpartacusOptions): string {
  const baseUrlPart = `baseUrl: '${options.baseUrl}',`;
  return `{
      backend: {
        occ: {${options.useMetaTags ? '' : baseUrlPart}
          prefix: '/rest/v2/'
        }
      },
      authentication: {
        client_id: 'mobile_android',
        client_secret: 'secret'
      },
      context: {
        baseSite: ['${options.baseSite}']
      },
      i18n: {
        resources: translations,
        chunks: translationChunksConfig,
        fallbackLang: 'en'
      },
      features: {
        level: '${options.featureLevel}'
      }
    }`;
}

function updateAppModule(options: SpartacusOptions): Rule {
  return (host: Tree, context: SchematicContext) => {
    context.logger.debug('Updating main module');

    // find app module
    const projectTargets = getProjectTargets(host, options.project);

    if (!projectTargets.build) {
      throw new SchematicsException(`Project target "build" not found.`);
    }

    const mainPath = projectTargets.build.options.main;
    const modulePath = getAppModulePath(host, mainPath);
    context.logger.debug(`main module path: ${modulePath}`);

    // add imports
    addImport(host, modulePath, 'translations', '@spartacus/assets');
    addImport(host, modulePath, 'translationChunksConfig', '@spartacus/assets');
    addImport(host, modulePath, 'B2cStorefrontModule', '@spartacus/storefront');
    addImport(
      host,
      modulePath,
      'defaultCmsContentConfig',
      '@spartacus/storefront'
    );
    addImport(host, modulePath, 'ConfigModule', '@spartacus/core');

    importModule(
      host,
      modulePath,
      `B2cStorefrontModule.withConfig(${getStorefrontConfig(options)})`
    );
    importModule(
      host,
      modulePath,
      `ConfigModule.withConfigFactory(defaultCmsContentConfig)`
    );

    return host;
  };
}

function installStyles(project: experimental.workspace.WorkspaceProject): Rule {
  return (host: Tree) => {
    const styleFilePath = getProjectStyleFile(project);

    if (!styleFilePath) {
      console.warn(
        red(`Could not find the default style file for this project.`)
      );
      console.warn(red(`Please consider manually setting up spartacus styles`));
      return;
    }

    if (styleFilePath.split('.').pop() !== 'scss') {
      console.warn(
        red(`Could not find the default SCSS style file for this project. `)
      );
      console.warn(
        red(
          `Please make sure your project is configured with SCSS and consider manually setting up spartacus styles.`
        )
      );
      return;
    }

    const buffer = host.read(styleFilePath);

    if (!buffer) {
      console.warn(
        red(
          `Could not read the default style file within the project ` +
            `(${italic(styleFilePath)})`
        )
      );
      console.warn(red(`Please consider manually importing spartacus styles.`));
      return;
    }

    const htmlContent = buffer.toString();
    const insertion = '\n' + `@import '~@spartacus/styles/index';\n`;

    if (htmlContent.includes(insertion)) {
      return;
    }

    const recorder = host.beginUpdate(styleFilePath);

    recorder.insertLeft(htmlContent.length, insertion);
    host.commitUpdate(recorder);
  };
}

function updateMainComponent(
  project: experimental.workspace.WorkspaceProject,
  options: SpartacusOptions
): Rule {
  return (host: Tree, _context: SchematicContext) => {
    const filePath = project.sourceRoot + '/app/app.component.html';
    const buffer = host.read(filePath);

    if (!buffer) {
      console.warn(red(`Could not read app.component.html file.`));
      return;
    }

    const htmlContent = buffer.toString();
    const insertion = '\n' + `<cx-storefront></cx-storefront>\n`;

    if (htmlContent.includes(insertion)) {
      return;
    }

    const recorder = host.beginUpdate(filePath);

    if (options && options.overwriteAppComponent) {
      recorder.remove(0, htmlContent.length);
      recorder.insertLeft(0, insertion);
    } else {
      recorder.insertLeft(htmlContent.length, insertion);
    }

    host.commitUpdate(recorder);

    return host;
  };
}

export function getIndexHtmlPath(
  project: experimental.workspace.WorkspaceProject
): string {
  const buildOptions = getProjectTargetOptions(project, 'build');

  if (!buildOptions.index) {
    throw new SchematicsException('"index.html" file not found.');
  }

  return buildOptions.index;
}

function updateIndexFile(
  project: experimental.workspace.WorkspaceProject,
  options: SpartacusOptions
): Rule {
  return (host: Tree) => {
    const projectIndexHtmlPath = getIndexHtmlPath(project);
    const baseUrl = options.baseUrl || 'OCC_BACKEND_BASE_URL_VALUE';

    const metaTags = [
      `<meta name="occ-backend-base-url" content="${baseUrl}" />`,
      `<meta name="media-backend-base-url" content="MEDIA_BACKEND_BASE_URL_VALUE" />`,
    ];

    metaTags.forEach(metaTag => {
      appendHtmlElementToHead(host, projectIndexHtmlPath, metaTag);
    });

    return host;
  };
}

export function addSpartacus(options: SpartacusOptions): Rule {
  return (tree: Tree, context: SchematicContext) => {
    const { workspace } = getWorkspace(tree);

    if (!options.project) {
      throw new SchematicsException('Option "project" is required.');
    }

    const project = workspace.projects[options.project];
    if (!project) {
      throw new SchematicsException(
        `Project is not defined in this workspace.`
      );
    }

    if (project.projectType !== 'application') {
      throw new SchematicsException(
        `Spartacus requires a project type of "application".`
      );
    }

    return chain([
      addPackageJsonDependencies(),
      updateAppModule(options),
      installStyles(project),
      updateMainComponent(project, options),
      options.useMetaTags ? updateIndexFile(project, options) : noop(),
      installPackageJsonDependencies(),
    ])(tree, context);
  };
}
