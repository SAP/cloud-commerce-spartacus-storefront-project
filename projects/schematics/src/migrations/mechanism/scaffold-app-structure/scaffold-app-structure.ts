import { Rule, Tree } from '@angular-devkit/schematics';
import {
  getDefaultProjectNameFromWorkspace,
  scaffoldStructure,
} from '../../../shared/utils/workspace-utils';

export function scaffoldAppStructure(tree: Tree): Rule {
  const project = getDefaultProjectNameFromWorkspace(tree);
  return scaffoldStructure({ project });
}
