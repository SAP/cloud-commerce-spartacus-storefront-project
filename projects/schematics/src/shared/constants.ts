export const UTF_8 = 'utf-8';

/***** Imports start *****/
export const ANGULAR_SCHEMATICS = '@schematics/angular';
export const ANGULAR_CORE = '@angular/core';
export const ANGULAR_FORMS = '@angular/forms';
export const ANGULAR_ROUTER = '@angular/router';
export const RXJS = 'rxjs';

export const SPARTACUS_CORE = '@spartacus/core';
export const SPARTACUS_STOREFRONTLIB = '@spartacus/storefront';
export const SPARTACUS_CDS = '@spartacus/cds';

export const NGRX_STORE = '@ngrx/store';
/***** Imports end *****/

/***** Classes start *****/
export const OBSERVABLE_CLASS = 'Observable';

export const STORE = 'Store';

export const FORM_BUILDER = 'FormBuilder';

export const CHANGE_DETECTOR_REF = 'ChangeDetectorRef';

export const ACTIVATED_ROUTE = 'ActivatedRoute';
export const ROUTER = 'Router';

export const CMS_COMPONENT_DATA_CLASS = 'CmsComponentData';
export const CONFIG_MODULE_CLASS = 'ConfigModule';
export const CMS_CONFIG = 'CmsConfig';
export const CMS_SELECTORS = 'CmsSelectors';
export const CMS_ACTIONS = 'CmsActions';
export const LOAD_CMS_COMPONENT_CLASS = 'LoadCmsComponent';
export const LOAD_CMS_COMPONENT_FAIL_CLASS = 'LoadCmsComponentFail';
export const LOAD_CMS_COMPONENT_SUCCESS_CLASS = 'LoadCmsComponentSuccess';
export const CMS_GET_COMPONENT_FROM_PAGE = 'CmsGetComponentFromPage';
export const USER_ADDRESS_SERVICE = 'UserAddressService';
export const AUTH_SERVICE = 'AuthService';
export const FEATURE_CONFIG_SERVICE = 'FeatureConfigService';
export const PAGE_META_RESOLVER = 'PageMetaResolver';
export const CMS_SERVICE = 'CmsService';
export const PAGE_META_SERVICE = 'PageMetaService';
export const CHECKOUT_SERVICE = 'CheckoutService';
export const CHECKOUT_PAYMENT_SERVICE = 'CheckoutPaymentService';
export const CHECKOUT_DELIVERY_SERVICE = 'CheckoutDeliveryService';
export const CART_DATA_SERVICE = 'CartDataService';
export const CART_SERVICE = 'CartService';
export const ACTIVE_CART_SERVICE = 'ActiveCartService';
export const PROMOTION_SERVICE = 'PromotionService';
export const ORDER_DETAILS_SERVICE = 'OrderDetailsService';
export const CHECKOUT_LOGIN_COMPONENT = 'CheckoutLoginComponent';
export const AUTH_REDIRECT_SERVICE = 'AuthRedirectService';
export const CHECKOUT_DETAILS_SERVICE = 'CheckoutDetailsService';
export const NOT_CHECKOUT_AUTH_GUARD = 'NotCheckoutAuthGuard';
export const ROUTING_SERVICE = 'RoutingService';
export const SHIPPING_ADDRESS_COMPONENT = 'ShippingAddressComponent';
export const CHECKOUT_CONFIG_SERVICE = 'CheckoutConfigService';
export const TRANSLATION_SERVICE = 'TranslationService';
export const CHECKOUT_PAGE_META_RESOLVER = 'CheckoutPageMetaResolver';
export const ADD_TO_CART_COMPONENT = 'AddToCartComponent';
export const MODAL_SERVICE = 'ModalService';
export const CURRENT_PRODUCT_SERVICE = 'CurrentProductService';
export const CART_NOT_EMPTY_GUARD = 'CartNotEmptyGuard';
export const CART_TOTALS_COMPONENT = 'CartTotalsComponent';
export const MINI_CART_COMPONENT = 'MiniCartComponent';
export const CHECKOUT_ORDER_SUMMARY_COMPONENT = 'CheckoutOrderSummaryComponent';
export const CHECKOUT_PROGRESS_MOBILE_TOP_COMPONENT =
  'CheckoutProgressMobileTopComponent';
export const CHECKOUT_CONFIG = 'CheckoutConfig';
export const ROUTING_CONFIG_SERVICE = 'RoutingConfigService';
export const PAYMENT_METHOD_COMPONENT = 'PaymentMethodComponent';
export const USER_PAYMENT_SERVICE = 'UserPaymentService';
export const GLOBAL_MESSAGE_SERVICE = 'GlobalMessageService';
export const CHECKOUT_AUTH_GUARD = 'CheckoutAuthGuard';
export const CART_PAGE_LAYOUT_HANDLER = 'CartPageLayoutHandler';
export const SELECTIVE_CART_SERVICE = 'SelectiveCartService';
export const CDS_SPARTACUS_EVENT_SERVICE = 'SpartacusEventService';
export const CONSENT_SERVICE = 'ConsentService';
export const CDS_CONFIG = 'CdsConfig';
export const ADDRESS_BOOK_COMPONENT_SERVICE = 'AddressBookComponentService';
export const CHECKOUT_GUARD = 'CheckoutGuard';
export const EXPRESS_CHECKOUT_SERVICE = 'ExpressCheckoutService';
export const CONSENT_MANAGEMENT_FORM_COMPONENT =
  'ConsentManagementFormComponent';
/***** Classes end *****/

/***** Removed public api start *****/
export const CART_EFFECTS = 'CartEffects';
export const WISHLIST_EFFECTS = 'WishlistEffects';
export const CART_VOUCHER_EFFECTS = 'CartVoucherEffects';
export const CART_ENTRY_EFFECTS = 'CartEntryEffects';
export const CART_COMBINED_EFFECTS = 'effects';

export const GET_REDUCERS = 'getReducers';
export const REDUCER_TOKEN = 'reducerToken';
export const REDUCER_PROVIDER = 'reducerProvider';
export const CLEAR_CART_STATE = 'clearCartState';
export const META_REDUCERS = 'metaReducers';
export const CLEAR_MULTI_CART_STATE = 'clearMultiCartState';
export const MULTI_CART_META_REDUCERS = 'multiCartMetaReducers';
export const MULTI_CART_REDUCER_TOKEN = 'multiCartReducerToken';
export const GET_MULTI_CART_REDUCERS = 'getMultiCartReducers';
export const MULTI_CART_REDUCER_PROVIDER = 'multiCartReducerProvider';

export const SITE_CONTEXT_PARAMS_PROVIDERS = 'siteContextParamsProviders';
export const INIT_SITE_CONTEXT_ROUTES_HANDLER = 'initSiteContextRoutesHandler';
export const INITITIALIZE_CONTEXT = 'inititializeContext';
export const CONTEXT_SERVICE_PROVIDERS = 'contextServiceProviders';
/***** Removed public api end *****/

/***** Properties start *****/
export const CMS_COMPONENT_DATA_PROPERTY_NAME = 'componentData';
/***** Properties end *****/

/***** APIs start *****/
export const TODO_SPARTACUS = 'TODO:Spartacus -';

export const GET_COMPONENT_STATE_OLD_API = 'getComponentState';
export const GET_COMPONENTS_STATE_NEW_API = 'getComponentsState';

export const GET_COMPONENT_ENTITIES_OLD_API = 'getComponentEntities';
export const GET_COMPONENT_ENTITIES_COMMENT = `// ${TODO_SPARTACUS} '${GET_COMPONENT_ENTITIES_OLD_API}' has been removed, please use some of the newer API methods.`;

export const COMPONENT_STATE_SELECTOR_FACTORY_OLD_API =
  'componentStateSelectorFactory';
export const COMPONENTS_STATE_SELECTOR_FACTORY_NEW_API =
  'componentsLoaderStateSelectorFactory';

export const COMPONENT_SELECTOR_FACTORY_OLD_API = 'componentSelectorFactory';
export const COMPONENTS_SELECTOR_FACTORY_NEW_API = 'componentsSelectorFactory';
/***** APIs end *****/

/***** Misc start *****/
export const CSS_DOCS_URL =
  'https://sap.github.io/cloud-commerce-spartacus-storefront-docs/deprecation-guide/css';
/***** Misc end *****/
