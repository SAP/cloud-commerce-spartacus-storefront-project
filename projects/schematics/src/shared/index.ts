export * from './constants';
export * from './feature-libs-constants';
export * from './utils/index';
