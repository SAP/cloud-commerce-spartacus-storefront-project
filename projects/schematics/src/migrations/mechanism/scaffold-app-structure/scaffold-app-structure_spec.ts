/// <reference types="jest" />

import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { Schema as SpartacusOptions } from '../../../add-spartacus/schema';
import { SPARTACUS_SCHEMATICS } from '../../../shared/constants';
import { scaffoldAppStructure } from './scaffold-app-structure';

const collectionPath = path.join(__dirname, '../../../collection.json');
const featuresModulePath = 'src/app/spartacus/spartacus-features.module.ts';
const configurationModulePath =
  'src/app/spartacus/spartacus-configuration.module.ts';
const spartacusModulePath = 'src/app/spartacus/spartacus.module.ts';

describe('scaffold app structure', () => {
  let appTree: UnitTestTree;
  let schematicRunner: SchematicTestRunner;

  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };

  const spartacusDefaultOptions: SpartacusOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [],
  };

  beforeEach(async () => {
    schematicRunner = new SchematicTestRunner(
      'test',
      require.resolve('../../test/migrations-test.json')
    );
    schematicRunner.registerCollection(SPARTACUS_SCHEMATICS, collectionPath);

    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
  });

  describe('When the new app structure does NOT exist', () => {
    it('should create it', async () => {
      const resultTree = await schematicRunner
        .callRule(scaffoldAppStructure, appTree)
        .toPromise();

      expect(resultTree.read(featuresModulePath)?.toString()).toMatchSnapshot();
      expect(
        resultTree.read(configurationModulePath)?.toString()
      ).toMatchSnapshot();
      expect(
        resultTree.read(spartacusModulePath)?.toString()
      ).toMatchSnapshot();
    });
  });

  describe('When the new app structure is already in place', () => {
    beforeEach(async () => {
      appTree = await schematicRunner
        .runExternalSchematicAsync(
          SPARTACUS_SCHEMATICS,
          'ng-add',
          { ...spartacusDefaultOptions, name: 'schematics-test' },
          appTree
        )
        .toPromise();
    });

    it('should not touch it', async () => {
      const resultTree = await schematicRunner
        .callRule(scaffoldAppStructure, appTree)
        .toPromise();

      expect(resultTree.read(featuresModulePath)?.toString()).toMatchSnapshot();
      expect(
        resultTree.read(configurationModulePath)?.toString()
      ).toMatchSnapshot();
      expect(
        resultTree.read(spartacusModulePath)?.toString()
      ).toMatchSnapshot();
    });
  });
});
