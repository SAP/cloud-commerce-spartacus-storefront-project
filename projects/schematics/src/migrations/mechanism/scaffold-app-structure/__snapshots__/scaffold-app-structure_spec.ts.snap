// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scaffold app structure When the new app structure does NOT exist should create it 1`] = `
"import { NgModule } from '@angular/core';



@NgModule({
  declarations: [],
  imports: [
  ]
})
export class SpartacusFeaturesModule { }
"
`;

exports[`scaffold app structure When the new app structure does NOT exist should create it 2`] = `
"import { NgModule } from '@angular/core';



@NgModule({
  declarations: [],
  imports: [
  ]
})
export class SpartacusConfigurationModule { }
"
`;

exports[`scaffold app structure When the new app structure does NOT exist should create it 3`] = `
"import { NgModule } from '@angular/core';
import { SpartacusFeaturesModule } from './spartacus-features.module';
import { SpartacusConfigurationModule } from './spartacus-configuration.module';



@NgModule({
  declarations: [],
  imports: [
  
    SpartacusFeaturesModule,
       SpartacusConfigurationModule
  ]
})
export class SpartacusModule { }
"
`;

exports[`scaffold app structure When the new app structure is already in place should not touch it 1`] = `
"import { NgModule } from '@angular/core';
import { AnonymousConsentsModule, AuthModule, CartModule, CartOccModule, CostCenterOccModule, ExternalRoutesModule, OrderOccModule, ProductModule, ProductOccModule, UserOccTransitionalModule, UserTransitionalModule } from \\"@spartacus/core\\";
import { AddressBookModule, AnonymousConsentManagementBannerModule, AnonymousConsentsDialogModule, BannerCarouselModule, BannerModule, BreadcrumbModule, CartComponentModule, CartPageEventModule, CategoryNavigationModule, CmsParagraphModule, ConsentManagementModule, FooterNavigationModule, HamburgerMenuModule, HomePageEventModule, LinkModule, LoginRouteModule, LogoutModule, MyCouponsModule, MyInterestsModule, NavigationEventModule, NavigationModule, NotificationPreferenceModule, OrderCancellationModule, OrderDetailsModule, OrderHistoryModule, OrderReturnModule, PaymentMethodsModule, ProductCarouselModule, ProductDetailsPageModule, ProductFacetNavigationModule, ProductImagesModule, ProductIntroModule, ProductListingPageModule, ProductListModule, ProductPageEventModule, ProductReferencesModule, ProductSummaryModule, ProductTabsModule, ReplenishmentOrderDetailsModule, ReplenishmentOrderHistoryModule, ReturnRequestDetailModule, ReturnRequestListModule, SearchBoxModule, SiteContextSelectorModule, StockNotificationModule, TabParagraphContainerModule, WishListModule } from \\"@spartacus/storefront\\";

@NgModule({
  declarations: [],
  imports: [
    // Auth Core
    AuthModule.forRoot(),
    LogoutModule,
    LoginRouteModule,
    // Basic Cms Components
    HamburgerMenuModule,
    SiteContextSelectorModule,
    LinkModule,
    BannerModule,
    CmsParagraphModule,
    TabParagraphContainerModule,
    BannerCarouselModule,
    CategoryNavigationModule,
    NavigationModule,
    FooterNavigationModule,
    BreadcrumbModule,
    // User Core,
    UserTransitionalModule,
    UserOccTransitionalModule,
    // User UI,
    AddressBookModule,
    PaymentMethodsModule,
    NotificationPreferenceModule,
    MyInterestsModule,
    StockNotificationModule,
    ConsentManagementModule,
    MyCouponsModule,
    // Anonymous Consents Core,
    AnonymousConsentsModule.forRoot(),
    // Anonymous Consents UI,
    AnonymousConsentsDialogModule,
    AnonymousConsentManagementBannerModule,
    // Product Core,
    ProductModule.forRoot(),
    ProductOccModule,
    // Product UI,
    ProductDetailsPageModule,
    ProductListingPageModule,
    ProductListModule,
    SearchBoxModule,
    ProductFacetNavigationModule,
    ProductTabsModule,
    ProductCarouselModule,
    ProductReferencesModule,
    ProductImagesModule,
    ProductSummaryModule,
    ProductIntroModule,
    // Cart Core,
    CartModule.forRoot(),
    CartOccModule,
    // Cart UI,
    CartComponentModule,
    WishListModule,
    CostCenterOccModule,
    // Order,
    OrderHistoryModule,
    OrderDetailsModule,
    OrderCancellationModule,
    OrderReturnModule,
    ReturnRequestListModule,
    ReturnRequestDetailModule,
    ReplenishmentOrderHistoryModule,
    ReplenishmentOrderDetailsModule,
    OrderOccModule,
    // Page Events,
    NavigationEventModule,
    HomePageEventModule,
    CartPageEventModule,
    ProductPageEventModule,
    // External routes,
    ExternalRoutesModule.forRoot(),
  ]
})
export class SpartacusFeaturesModule { }
"
`;

exports[`scaffold app structure When the new app structure is already in place should not touch it 2`] = `
"import { NgModule } from '@angular/core';
import { translationChunksConfig, translations } from \\"@spartacus/assets\\";
import { FeaturesConfig, I18nConfig, OccConfig, provideConfig, SiteContextConfig } from \\"@spartacus/core\\";
import { defaultCmsContentProviders, layoutConfig, mediaConfig } from \\"@spartacus/storefront\\";

@NgModule({
  declarations: [],
  imports: [
  ],
  providers: [provideConfig(layoutConfig), provideConfig(mediaConfig), ...defaultCmsContentProviders, provideConfig(<OccConfig>{
    backend: {
      occ: {
        baseUrl: 'https://localhost:9002',
      }
    },
  }), provideConfig(<SiteContextConfig>{
    context: {
      currency: ['USD'],
      language: ['en'],
    },
  }), provideConfig(<I18nConfig>{
    i18n: {
      resources: translations,
      chunks: translationChunksConfig,
      fallbackLang: 'en'
    },
  }), provideConfig(<FeaturesConfig>{
    features: {
      level: '3.2'
    }
  })]
})
export class SpartacusConfigurationModule { }
"
`;

exports[`scaffold app structure When the new app structure is already in place should not touch it 3`] = `
"import { NgModule } from '@angular/core';
import { BaseStorefrontModule } from \\"@spartacus/storefront\\";
import { SpartacusConfigurationModule } from './spartacus-configuration.module';
import { SpartacusFeaturesModule } from './spartacus-features.module';

@NgModule({
  declarations: [],
  imports: [

    SpartacusFeaturesModule,
    SpartacusConfigurationModule,
    BaseStorefrontModule
  ],
  exports: [BaseStorefrontModule]
})
export class SpartacusModule { }
"
`;
